File: app/src/main/java/com/example/waveplayer/RecyclerViewAdapterPlaylists.java
package com.example.waveplayer;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class RecyclerViewAdapterPlaylists extends RecyclerView.Adapter<RecyclerViewAdapterPlaylists.ViewHolder> {

    private final Fragment fragment;

    public List<RandomPlaylist> randomPlaylists;

    OnCreateContextMenuListenerPlaylists onCreateContextMenuListenerPlaylists;

    View.OnClickListener onClickListenerHandle;

    View.OnClickListener onClickListenerViewHolder;

    public RecyclerViewAdapterPlaylists(Fragment fragment, List<RandomPlaylist> items) {
        this.fragment = fragment;
        randomPlaylists = items;
    }

    public void updateList(List<RandomPlaylist> randomPlaylists) {
        this.randomPlaylists = randomPlaylists;
        notifyDataSetChanged();
    }

    @Override
    @NonNull
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_playlist, parent, false));
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.randomPlaylist = randomPlaylists.get(position);
        holder.textViewPlaylistName.setText(randomPlaylists.get(position).getName());
        holder.handle.setOnCreateContextMenuListener(null);
        onCreateContextMenuListenerPlaylists =
                new OnCreateContextMenuListenerPlaylists(fragment, holder.randomPlaylist);
        holder.handle.setOnCreateContextMenuListener(onCreateContextMenuListenerPlaylists);
        onClickListenerHandle = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                holder.handle.performLongClick();
            }
        };
        holder.handle.setOnClickListener(null);
        holder.handle.setOnClickListener(onClickListenerHandle);
        onClickListenerViewHolder = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int position = holder.getAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    ActivityMain activityMain = ((ActivityMain) fragment.getActivity());
                    if (activityMain != null) {
                        activityMain.serviceMain.userPickedPlaylist = holder.randomPlaylist;
                    }
                    NavHostFragment.findNavController(fragment).navigate(
                            FragmentPlaylistsDirections.actionFragmentPlaylistsToFragmentPlaylist());
                }
            }
        };
        holder.playlistView.setOnClickListener(null);
        holder.playlistView.setOnClickListener(onClickListenerViewHolder);
    }

    @Override
    public void onViewRecycled(@NonNull ViewHolder holder) {
        super.onViewRecycled(holder);
        onCreateContextMenuListenerPlaylists = null;
        holder.handle.setOnCreateContextMenuListener(null);
        onClickListenerHandle = null;
        holder.handle.setOnClickListener(null);
        onClickListenerViewHolder = null;
        holder.playlistView.setOnClickListener(null);
        holder.playlistView = null;
        holder.textViewPlaylistName = null;
        holder.randomPlaylist = null;
    }

    @Override
    public int getItemCount() {
        return randomPlaylists.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        public View playlistView;
        public TextView textViewPlaylistName;
        public RandomPlaylist randomPlaylist;
        final ImageView handle;

        public ViewHolder(View view) {
            super(view);
            playlistView = view;
            textViewPlaylistName = view.findViewById(R.id.text_view_playlist_name);
            if (randomPlaylist != null) {
                textViewPlaylistName.setText(randomPlaylist.getName());
            }
            handle = view.findViewById(R.id.playlist_handle);
        }

        @Override
        @NonNull
        public String toString() {
            return randomPlaylist.getName();
        }

    }

}

File: app/src/main/java/com/fourthFinger/pinkyPlayer/fragments/RecyclerViewAdapterPlaylists.kt
package com.fourthFinger.pinkyPlayer.fragments

import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.fourthFinger.pinkyPlayer.DiffUtils
import com.fourthFinger.pinkyPlayer.R
import com.fourthFinger.pinkyPlayer.random_playlist.RandomPlaylist

class RecyclerViewAdapterPlaylists(
        private var listenerCallbackPlaylists: ListenerCallbackPlaylists,
        randomPlaylists: List<RandomPlaylist>
) : ListAdapter<RandomPlaylist, RecyclerViewAdapterPlaylists.ViewHolder>(
    DiffUtils.DiffUtilItemCallbackPlaylists()
) {

    interface ListenerCallbackPlaylists {
        fun onClickViewHolder(randomPlaylist: RandomPlaylist)
        fun onMenuItemClickAddToPlaylist(randomPlaylist: RandomPlaylist): Boolean
        fun onMenuItemClickAddToQueue(randomPlaylist: RandomPlaylist): Boolean
    }

    fun updateList(randomPlaylists: List<RandomPlaylist>) {
        submitList(ArrayList(randomPlaylists))
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(
            R.layout.item_playlist,
            parent,
            false)
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.randomPlaylist = getItem(position)
        holder.textViewPlaylistName.text = getItem(position).getName()
        holder.handle.setOnCreateContextMenuListener{ menu: ContextMenu?, _: View?, _: ContextMenuInfo? ->
            val itemAddToPlaylist = menu?.add(R.string.add_to_playlist)
            itemAddToPlaylist?.setOnMenuItemClickListener{
                listenerCallbackPlaylists.onMenuItemClickAddToPlaylist(getItem(position))
            }
            val itemAddToQueue = menu?.add(R.string.add_to_queue)
            itemAddToQueue?.setOnMenuItemClickListener{
                listenerCallbackPlaylists.onMenuItemClickAddToQueue(getItem(position))
            }
        }
        holder.handle.setOnClickListener{ holder.handle.performLongClick() }
        holder.playlistView.setOnClickListener{
            if (position != RecyclerView.NO_POSITION) {
                listenerCallbackPlaylists.onClickViewHolder(getItem(position))
            }
        }
    }

    init {
        updateList(randomPlaylists)
    }

    class ViewHolder(val playlistView: View) : RecyclerView.ViewHolder(playlistView) {
        val textViewPlaylistName: TextView = playlistView.findViewById(R.id.text_view_playlist_name)
        val handle: ImageView = playlistView.findViewById(R.id.playlist_handle)

        lateinit var randomPlaylist: RandomPlaylist

        override fun toString(): String {
            return randomPlaylist.getName()
        }

    }

}